### PR TITLE
Fix parsing of tokenized strings with meta characters

### PR DESCRIPTION
**Pull request by adampauls**
Removes special parsing of tokenized Lispress, since Lispress parsing already whitespace-invariant enough.

### CODE REVIEW

File: tests/test_dataflow/core/test_linearize.py [hunk 1]
```diff
@@ -139,3 +140,11 @@ def test_meta():
         ")",
         ")",
     ]
+
+
+def test_meta_to_canonical():
+    s = """( Yield ( Execute ( ReviseConstraint ( refer ( ^ ( Dynamic ) roleConstraint ( Path.apply "output" ) ) ) ( ^ ( Event ) ConstraintTypeIntension ) ( Event.showAs_? ( ?= ( ShowAsStatus.OutOfOffice ) ) ) ) ) )"""
```
Comment (sammthomson) [inline tests/test_dataflow/core/test_linearize.py:146]: It was my understanding that `"output"` would actually look like `" output "` in the linearized output, and that this was important to getting good numbers with seq2seq. @hao-fang is that true?

Comment (sammthomson) [reply tests/test_dataflow/core/test_linearize.py:146]: maybe `"output"` is not the best example for it mattering... more important are strings that were copied from the utterance, like `# ( PersonName " Dan Schoffel " )`

Comment (hao-fang) [reply tests/test_dataflow/core/test_linearize.py:146]: Yeah, you are right.
But if there is no space in the double quote for `Path.apply` then it may be okay?
The `# ( PersonName " Dan Schoffel " )` example may be handled by the `sexp_formatter` and `sexp_deformatter`.

Comment (adampauls) [reply tests/test_dataflow/core/test_linearize.py:146]: Yeah, sexp_deformatter handles it. 

I'll add a test case. This is a copy-paste from Elias so I don't know why the quotes look that.

File: src/dataflow/core/linearize.py [hunk 2]
```diff
@@ -58,17 +59,6 @@ def sexp_deformatter(sexp_tokens: List[str]) -> List[str]:
     """
     sexp_str = " ".join(sexp_tokens)
 
-    sexp_str = re.sub(
```
Comment (adampauls) [inline src/dataflow/core/linearize.py:61]: I think we might never have had real JSON anyways.

File: src/dataflow/core/linearize.py [hunk 3]
```diff
@@ -58,17 +59,6 @@ def sexp_deformatter(sexp_tokens: List[str]) -> List[str]:
...
-        r'{ "schema" : "(\S+)" , "underlying" : " ([\S+\s*]+?) " }',
-        r'{ "schema" : "\1" , "underlying" : "\2" }',
-        sexp_str,
-    )
-
-    sexp_str = re.sub(
```
Comment (adampauls) [inline src/dataflow/core/linearize.py:67]: This happens on the parsed sexp now.

File: src/dataflow/core/lispress.py [hunk 4]
```diff
@@ -80,18 +80,19 @@ def normalize_numbers(exp: Lispress) -> "Lispress":
...
-                return '"' + exp[1:-1].strip() + '"'
-            else:
-                return exp
-        else:
-            return [strip_copy_strings(e) for e in exp]
-
     return render_compact(strip_copy_strings(normalize_numbers(round_tripped)))
 
 
+def strip_copy_strings(exp: Lispress) -> "Lispress":
```
Comment (adampauls) [inline src/dataflow/core/lispress.py:86]: Make public.

File: tests/test_dataflow/core/test_linearize.py [hunk 5]
```diff
@@ -97,31 +96,22 @@ def test_unnest_line():
 def test_sexp_formatter_and_deformatter():
     """Round-trip tests for s-expression formatter and deformatter."""
     data = [
-        ("", ""),
-        ('# ( String "singleToken" )', '# ( String " singleToken " )'),
-        ('# ( String "multiple tokens" )', '# (  String " multiple tokens " )'),
-        # these two are not valid value s-expressions, so they are not formatted
-        ('# ( String "open quote only )', '# (  String "open quote only )'),
-        ('# ( String closing quote only" )', '# ( String closing quote only" )',),
-        # multiple value s-expressions
```
Comment (adampauls) [inline tests/test_dataflow/core/test_linearize.py:106]: Not sure why were were testing these things.

**Review by sammthomson — APPROVED** [review]
LGTM 👍

File: tests/test_dataflow/core/test_linearize.py (hunk 5 above)
Comment (hao-fang) [reply tests/test_dataflow/core/test_linearize.py:106]: Are they broken with the new changes? If not, maybe we can just keep them.
I think these were intended to test corner cases. They may have appeared in some version of the data before, although not the current versions.

File: tests/test_dataflow/core/test_linearize.py [hunk 6]
```diff
@@ -97,31 +96,22 @@ def test_unnest_line():
...
             "( Constraint[Constraint[flight]] ) ( "
-            'Constraint :type ( ?= ( # ( String " fare " ) ) ) ) ) ) )',
+            'Constraint :type ( ?= # ( String " fare " ) ) ) ) ) )',
         ),
     ]
 
     for raw_sexp, formatted_sexp in data:
-        assert sexp_formatter(raw_sexp.split()) == formatted_sexp.split()
-        assert sexp_deformatter(formatted_sexp.split()) == raw_sexp.split()
+        assert lispress_to_seq(parse_lispress(raw_sexp)) == formatted_sexp.split()
```
Comment (hao-fang) [inline tests/test_dataflow/core/test_linearize.py:113]: sorry I'm a little confused now. This test `test_sexp_formatter_and_deformatter` specifically checks the `sexp_formatter` and `sexp_deformatter`. Why we remove the `sexp_deformatter` call now?

File: src/dataflow/core/linearize.py (hunk 2 above)
Comment (hao-fang) [reply src/dataflow/core/linearize.py:61]: to make sure I'm on the same page -- it seems that we do not care backward compatibility, i.e., the code would broken on the SMCalFlow 1.0 data?
If so, probably we should change the README.md since the instructions there would be broken.

File: tests/test_dataflow/core/test_linearize.py (hunk 6 above)
Comment (adampauls) [reply tests/test_dataflow/core/test_linearize.py:113]: I think we were testing implementation details. Do you think it's important to test the formatter itself, or just that roundtripping works and that formatting works as expected? I can rename the test.

File: src/dataflow/core/linearize.py (hunk 2 above)
Comment (adampauls) [reply src/dataflow/core/linearize.py:61]: We do, I think all the changes I'm making can read in the old format. There is no JSON in CalFlow. 
```(base) jajacMScomp:metals adpauls$ grep "schema" ~/sm/calflow/train.dataflow_dialogues.jsonl
(base) jajacMScomp:metals adpauls$ grep "underlying" ~/sm/calflow/train.dataflow_dialogues.jsonl```

File: tests/test_dataflow/core/test_linearize.py (hunk 5 above)
Comment (adampauls) [reply tests/test_dataflow/core/test_linearize.py:106]: Yeah, because the formatting operates on parsed s-expressions and these are not valid s-expressions. I don't think we had any invalid s-expressions in the old data?

File: src/dataflow/core/linearize.py (hunk 2 above)
Comment (hao-fang) [reply src/dataflow/core/linearize.py:61]: I see. Maybe there are some dead code [here](https://github.com/microsoft/task_oriented_dialogue_as_dataflow_synthesis/blob/ab0a6d65ead8edfea1624bf0051ccab72c394f51/src/dataflow/core/lispress.py#L267).

It seems that removing the `re.sub`, the `sexp_deformatter` is trivial. Maybe we do not need this method anymore?
Can you also remove the `re.sub` in `sexp_formatter` (or completely delete these two methods)?

File: tests/test_dataflow/core/test_linearize.py (hunk 5 above)
Comment (hao-fang) [reply tests/test_dataflow/core/test_linearize.py:106]: Got it. Then removing them is fine.

File: tests/test_dataflow/core/test_linearize.py (hunk 6 above)
Comment (hao-fang) [reply tests/test_dataflow/core/test_linearize.py:113]: Yeah, I think it's better to rename the test. Thanks!

**Review by hao-fang — APPROVED** [review]
Thanks! LGTM modulo maybe deleting `sexp_formatter` and `sexp_deformatter`.

File: tests/test_dataflow/core/test_linearize.py (hunk 6 above)
Comment (adampauls) [reply tests/test_dataflow/core/test_linearize.py:59]: Done
